lumen/ai/agents.py

```diff
@@ -712,8 +714,54 @@ async def _render_sql_result(query, active):
         self.interface.stream(tabs, user="SQL", replace=True)
         tabs.active = 1
 
-    async def answer(self, messages: list | str):
+    @retry_llm_output()
+    async def _create_valid_sql(self, messages, system, source, tables, errors=None):
+        if errors:
+            last_query = self.interface.objects[-1].object.replace("```sql", "").rstrip("```").strip()
+            errors = '\n'.join(errors)
+            messages += [
+                {
+                    "role": "user",
+                    "content": (
+                        f"Your last query `{last_query}` did not work as intended, "
+                        f"expertly revise, and please do not repeat these issues:\n{errors}")
+                }
+            ]
+        print(messages)
+
         message = None
+        async for chunk in self.llm.stream(
+            messages,
+            system=system,
+            response_model=Sql,
+            field="query",
+        ):
+            if chunk is None:
+                continue
+            message = self.interface.stream(
+                f"```sql\n{chunk}\n```",
+                user="SQL",
+                message=message,
+                replace=True,
+            )
+        if message.object is None:
+            return
+        sql_query = message.object.replace("```sql", "").replace("```", "").strip()
+
+        # check whether the SQL query is valid
+        try:
+            transforms = [SQLOverride(override=sql_query), SQLLimit(limit=1)]
+            pipeline = Pipeline(
+                source=source, table=tables[0], sql_transforms=transforms
+            )
+            pipeline.data
```

Comment:
This validates the SQL (ensuring it's runnable) before the rendering, so that the LLM can auto-correct itself.